includes/Core/Util/Migration_1_128_0.php

```diff
@@ -0,0 +1,187 @@
+<?php
+/**
+ * Migration for 1.128.0
+ *
+ * @package   Google\Site_Kit\Core\Util
+ * @copyright 2024 Google LLC
+ * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ * @link      https://sitekit.withgoogle.com
+ */
+
+namespace Google\Site_Kit\Core\Util;
+
+use Google\Site_Kit\Context;
+use Google\Site_Kit\Core\Modules\Modules;
+use Google\Site_Kit\Core\Storage\Options;
+use Google\Site_Kit\Modules\Analytics_4\Settings as Analytics_Settings;
+use Google\Site_Kit\Modules\Ads;
+use Google\Site_Kit\Modules\Ads\Settings as Ads_Settings;
+
+/**
+ * Class Migration_1_128_0
+ *
+ * @since 1.128.0
+ * @access private
+ * @ignore
+ */
+class Migration_1_128_0 {
+	/**
+	 * Target DB version.
+	 */
+	const DB_VERSION = '1.128.0';
+
+	/**
+	 * DB version option name.
+	 */
+	const DB_VERSION_OPTION = 'googlesitekit_db_version';
+
+	/**
+	 * Analytics options name.
+	 */
+	const ANALYTICS_OPTIONS = 'googlesitekit_analytics-4_settings';
+
+	/**
+	 * Ads options name.
+	 */
+	const ADS_OPTIONS = 'googlesitekit_ads_settings';
+
+	/**
+	 * Context instance.
+	 *
+	 * @since 1.128.0
+	 * @var Context
+	 */
+	protected $context;
+
+	/**
+	 * Options instance.
+	 *
+	 * @since 1.128.0
+	 * @var Options
+	 */
+	protected $options;
+
+	/**
+	 * Analytics_Settings instance.
+	 *
+	 * @since 1.128.0
+	 * @var Analytics_Settings
+	 */
+	protected $analytics_settings;
+
+	/**
+	 * Ads_Settings instance.
+	 *
+	 * @since 1.128.0
+	 * @var Ads_Settings
+	 */
+	protected $ads_settings;
+
+	/**
+	 * Constructor.
+	 *
+	 * @since 1.128.0
+	 *
+	 * @param Context $context Plugin context instance.
+	 * @param Options $options Optional. Options instance.
+	 */
+	public function __construct(
+		Context $context,
+		Options $options = null
+	) {
+		$this->context            = $context;
+		$this->options            = $options ?: new Options( $context );
+		$this->analytics_settings = new Analytics_Settings( $this->options );
+		$this->ads_settings       = new Ads_Settings( $this->options );
+	}
+
+	/**
+	 * Registers hooks.
+	 *
+	 * @since 1.128.0
+	 */
+	public function register() {
+		add_action( 'admin_init', array( $this, 'migrate' ) );
+	}
+
+	/**
+	 * Migrates the DB.
+	 *
+	 * @since 1.128.0
+	 */
+	public function migrate() {
+		$db_version = $this->options->get( self::DB_VERSION_OPTION );
+
+		if ( ! $db_version || version_compare( $db_version, self::DB_VERSION, '<' ) ) {
+			$this->migrate_analytics_conversion_id_setting();
+			$this->activate_ads_module();
```

Comment:
Why do you activate the ads module? If that module hasn't been activated yet, it should remain disconnected even after the migration. If it has been connected, then there is no sense to re-activate it. Let's remove this.

Reply:
@eugene-manuilov one key purpose of this ticket is to move the Ads Conversion ID migration from the frontend to the backend. We need to activate the Ads module automatically if the user has a conversion ID so that the site will continue to track data with the ads tag. You can see in the [testing notes](https://docs.google.com/document/d/1D2zd0-HFfVU0Rx5Nq5MbEx_hwU-_sMxeSFkwx8HSgws/edit?disco=AAABOjo4lYs) the Ads module should be activated as part of the migration.

Reply:
Do you mean, skip activating it if it's already active?